README.md
# Old LZO Ports

This is an unofficial copy of the original `java-lzo`, `perl-lzo` and `python-lzo` libraries by Markus F. X. J. Oberhumer, released under the GNU General Public License version 2 (see the `COPYING` files in the individual directories). These ports of the LZO library to Java, Perl and Python were originally released in the years 1998-1999 and are provided here mostly for historical purposes, as they are no longer available on [the LZO page](http://www.oberhumer.com/opensource/lzo/) on http://www.oberhumer.com/.

`java-lzo` 1.00 was written for JDK 1.0 and 1.1, but it still appears to work under Java 8. A newer, unrelated implementation of LZO in Java, written by Karmasphere/Shevek and others, can be found [on GitHub](https://github.com/Karmasphere/lzo-java).

An official updated version of `perl-lzo`/`Compress::LZO` can be found [on CPAN](https://metacpan.org/pod/Compress::LZO).

`python-lzo` 1.00 was written for Python 1.5. An updated version, by Joshua D. Boyd and others, can be found [on PyPI](https://pypi.python.org/pypi/python-lzo) with source code [on GitHub](https://github.com/jd-boyd/python-lzo).


java-lzo/NEWS
============================================================================
User visible changes for Java-LZO
============================================================================

Changes in 1.00 (01 Dec 1999)
  * first public release



python-lzo/test.py
#! /usr/bin/env python
##
## vi:ts=4:et
##
##---------------------------------------------------------------------------##
##
## This file is part of the LZO real-time data compression library.
##
## Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer
##
## The LZO library is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License as
## published by the Free Software Foundation; either version 2 of
## the License, or (at your option) any later version.
##
## The LZO library is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with the LZO library; see the file COPYING.
## If not, write to the Free Software Foundation, Inc.,
## 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
##
## Markus F.X.J. Oberhumer
## markus.oberhumer@jk.uni-linz.ac.at
##
##---------------------------------------------------------------------------##

import sys, string
import lzo


# /***********************************************************************
# // a very simple test driver...
# ************************************************************************/

def print_modinfo():
    #print sys.modules
    mod = sys.modules['lzo']
    #print mod
    d = mod.__dict__
    for k in d.keys():
        print k, d[k]


def test(src, level = 1):
    a0 = lzo.adler32(src)
    c =  lzo.compress(src,level)
    u1 = lzo.decompress(c)
    a1 = lzo.adler32(u1)
    o =  lzo.optimize(c)
    u2 = lzo.decompress(o)
    a2 = lzo.adler32(u2)
    if cmp(src,u1) != 0 or cmp(src,u2) != 0:
        raise lzo.error, "internal error 1"
    if cmp(a0,a1) != 0 or cmp(a0,a2) != 0:
        raise lzo.error, "internal error 2"
    print "compressed %6d -> %6d" % (len(src), len(c))


def main(args):
    ## print_modinfo()

    # print version information and module documentation
    print "LZO version %s (0x%x), %s" % (lzo.LZO_VERSION_STRING, lzo.LZO_VERSION, lzo.LZO_VERSION_DATE)
    print
    print lzo.__doc__

    # compress some simple strings
    test("aaaaaaaaaaaaaaaaaaaaaaaa")
    test("abcabcabcabcabcabcabcabc")
    test("abcabcabcabcabcabcabcabc",9)
    test(" " * 131072)
    test("")
    print "Simple compression test passed."

    # force an exception
    try:
        x = lzo.decompress("xx")
    except lzo.error, msg:
        ## print msg
        pass
    else:
        print "Exception handling does NOT work !"


if __name__ == '__main__':
    main(sys.argv)



perl-lzo/LZO.pm
#  LZO.pm --
#
#  This file is part of the LZO real-time data compression library.
#
#  Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer
#
#  The LZO library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU General Public License as
#  published by the Free Software Foundation; either version 2 of
#  the License, or (at your option) any later version.
#
#  The LZO library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with the LZO library; see the file COPYING.
#  If not, write to the Free Software Foundation, Inc.,
#  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
#  Markus F.X.J. Oberhumer
#  markus.oberhumer@jk.uni-linz.ac.at


package Compress::LZO;

require 5.003_05 ;
require Exporter;
require DynaLoader;
use AutoLoader;
use Carp;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK $AUTOLOAD);

@ISA = qw(Exporter DynaLoader);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.
@EXPORT = qw(
	LZO_VERSION LZO_VERSION_STRING LZO_VERSION_DATE
);

@EXPORT_OK = qw(
	compress decompress optimize
	adler32 crc32
	LZO_VERSION LZO_VERSION_STRING LZO_VERSION_DATE
);


$VERSION = "1.00";


sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.  If a constant is not found then control is passed
    # to the AUTOLOAD in AutoLoader.
    my($constname);
    ($constname = $AUTOLOAD) =~ s/.*:://;
    my $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	if ($! =~ /Invalid/) {
	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
	    goto &AutoLoader::AUTOLOAD;
	}
	else {
	    croak "Compress::LZO macro $constname not defined";
	}
    }
    eval "sub $AUTOLOAD { $val }";
    goto &$AUTOLOAD;
}

bootstrap Compress::LZO $VERSION;

# Preloaded methods go here.


1 ;
# Autoload methods go after __END__, and are processed by the autosplit program.

1;
__END__


=cut

=head1 NAME

Compress::LZO - Interface to LZO compression library

=head1 SYNOPSIS

    use Compress::LZO;

    $dest = Compress::LZO::compress($source, [$level]);
    $dest = Compress::LZO::decompress($source);
    $dest = Compress::LZO::optimize($source);

    $crc = Compress::LZO::adler32($buffer [,$crc]);
    $crc = Compress::LZO::crc32($buffer [,$crc]);

    LZO_VERSION, LZO_VERSION_STRING, LZO_VERSION_DATE

=head1 DESCRIPTION

The I<Compress::LZO> module provides a Perl interface to the I<LZO>
compression library (see L</AUTHOR> for details about where to get
I<LZO>). A relevant subset of the functionality provided by I<LZO>
is available in I<Compress::LZO>.

All string parameters can either be a scalar or a scalar reference.

=head1 COMPRESSION FUNCTIONS

$dest = Compress::LZO::compress($string)

Compress a string using the default compression level, returning a string
containing compressed data.

$dest = Compress::LZO::compress($string, $level)

Compress string, using the chosen compression level (either 1 or 9).
Return a string containing the compressed data.

On error I<undef> is returned.

=head1 DECOMPRESSION FUNCTIONS

$dest = Compress::LZO::decompress($string)

Decompress the data in string, returning a string containing the
decompressed data.

On error I<undef> is returned.

=head1 OPTIMIZATION FUNCTIONS

$dest = Compress::LZO::optimize($string)

Optimize the representation of the compressed data, returning a
string containing the compressed data.

On error I<undef> is returned.

=head1 CHECKSUM FUNCTIONS

Two functions are provided by I<LZO> to calculate a checksum. For the
Perl interface, the order of the two parameters in both functions has
been reversed. This allows both running checksums and one off
calculations to be done.

    $crc = Compress::LZO::adler32($string [,$crc]) ;
    $crc = Compress::LZO::crc32($string [,$crc]) ;

=head1 AUTHOR

The I<Compress::LZO> module was written by Markus F.X.J. Oberhumer
F<markus.oberhumer@jk.uni-linz.ac.at>.
The latest copy of the module should also be found on CPAN in
F<modules/by-module/Compress/perl-lzo-x.x.tar.gz>.

The I<LZO> compression library was written by Markus F.X.J. Oberhumer
F<markus.oberhumer@jk.uni-linz.ac.at>.
It is available from the LZO home page at
F<http://wildsau.idv.uni-linz.ac.at/mfx/lzo.html>.

=head1 MODIFICATION HISTORY

1.00 First public release of I<Compress::LZO>.



java-lzo/lunpack.java
/* lunpack.java -- example program

   This file is part of the LZO real-time data compression library.

   Copyright (C) 1999 Markus Franz Xaver Johannes Oberhumer
   Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer
   Copyright (C) 1997 Markus Franz Xaver Johannes Oberhumer
   Copyright (C) 1996 Markus Franz Xaver Johannes Oberhumer

   The LZO library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   The LZO library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the LZO library; see the file COPYING.
   If not, write to the Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Markus F.X.J. Oberhumer
   <markus.oberhumer@jk.uni-linz.ac.at>
   http://wildsau.idv.uni-linz.ac.at/mfx/lzo.html
 */


import org.lzo.*;
import java.io.*;
// import java.util.zip.Adler32;  // @JDK@ 1.1


/***********************************************************************
 * This program serves as a test driver for the LZO1X decompressor.
 *
 * It can test and unpack files that were created with the
 * lpack example program from LZO 1.06.
 *
 * @author  Markus F.X.J. Oberhumer <markus.oberhumer@jk.uni-linz.ac.at>
 * @see     examples/lpack.c in the LZO 1.06 distribution
 ***********************************************************************/

class lunpack
{
    Lzo1xDecompressor bc = null;

    boolean opt_checksum = true;
    boolean opt_debug = false;

    // magic file header for compressed lpack files
    static final byte magic[] =
        { 0x00, (byte)0xe9, 0x4c, 0x5a, 0x4f, (byte)0xff, 0x1a };

    // for statistics
    int total_in = 0;
    int total_out = 0;
    int total_blocks = 0;
    int total_c = 0;
    int total_d = 0;

    // for benchmarks
    int compress_loops = 1;
    int decompress_loops = 1;
    long compress_time = 0;
    long decompress_time = 0;
    long total_time = 0;


    //
    // util
    //

    int xread(InputStream f, byte buf[], int off, int len, boolean allow_eof)
        throws IOException
    {
        int l = Util.xread(f,buf,off,len,allow_eof);
        total_in += l;
        return l;
    }

    void xwrite(OutputStream f, byte buf[], int off, int len)
        throws IOException
    {
        Util.xwrite(f,buf,off,len);
        total_out += len;
    }

    int xread32(InputStream f)
        throws IOException
    {
        int v = Util.xread32(f);
        total_in += 4;
        return v;
    }

    void xwrite32(OutputStream f, int v)
        throws IOException
    {
        Util.xwrite32(f,v);
        total_out += 4;
    }

    int xgetc(InputStream f)
        throws IOException
    {
        int v = Util.xgetc(f);
        total_in += 1;
        return v;
    }

    void xputc(OutputStream f, int v)
        throws IOException
    {
        Util.xputc(f,v);
        total_out += 1;
    }


    //
    // decompression
    //

    void do_decompress(InputStream fi, OutputStream fo)
        throws Exception
    {
        reset_stats();
        long total_start_time = System.currentTimeMillis();

        /*
         * Step 1: check magic header, read flags & block size, init checksum
         */
        byte m[] = new byte[magic.length];
        if (xread(fi,m,0,m.length,true) != m.length ||
            Util.memcmp(m,magic,m.length) != 0) {
            throw new DataFormatException("header error - this file is not compressed by lpack");
        }
        int flags = xread32(fi);
        int method = xgetc(fi);
        int level = xgetc(fi);
        if (method != 1) {
            throw new DataFormatException("header error - invalid method " + method);
        }
        int block_size = xread32(fi);
        if (block_size < 1024 || block_size > 1024*1024) {
            throw new DataFormatException("header error - invalid block size " + block_size);
        }
        Adler32 checksum = null;
        if (opt_checksum && (flags & 1) != 0)
            checksum = new Adler32();

        /*
         * Step 2: allocate buffers for decompression
         */
        int buf_len;
        byte in_buf[] = null;
        byte out_buf[] = null;

        if (decompress_loops == 1) {
            // we use in-place decompression to save some memory
            buf_len = block_size + block_size / 64 + 16 + 3;
            in_buf = new byte[buf_len];
            out_buf = in_buf;
        }
        else
        {
            // cannot use in-place decompression when running benchmark
            buf_len = block_size;
            in_buf = new byte[buf_len];
            out_buf = new byte[buf_len];
        }

        /*
         * Step 3: process blocks
         */
        for (;;)
        {
            int in;
            int out;
            int in_len;
            int out_len;

            /* read uncompressed size */
            out_len = xread32(fi);

            /* exit if last block (EOF marker) */
            if (out_len == 0)
                break;

            /* read compressed size */
            in_len = xread32(fi);

            /* sanity check of the size values */
            if (in_len > block_size || out_len > block_size ||
                in_len <= 0 || out_len <= 0 || in_len > out_len) {
                throw new DataFormatException("block size error - data corrupted");
            }

            /* place compressed block at the top of in_buf[] */
            in = buf_len - in_len;
            out = 0;
            xread(fi,in_buf,in,in_len,false);

            if (in_len < out_len)
            {
                /* decompress */
                int loops = decompress_loops;
                long start_time = System.currentTimeMillis();
                do {
                    Int new_len = new Int(out_len);
                    int r = bc.decompress(in_buf,in,in_len,out_buf,out,new_len);
                    if (r != bc.LZO_E_OK || new_len.intValue() != out_len) {
                        //System.out.println(r + " " + in_len + " " + out_len + " " + new_len.intValue());
                        throw new DataFormatException("compressed data violation");
                    }
                } while (--loops > 0);
                decompress_time += System.currentTimeMillis() - start_time;
                /* write decompressed block */
                xwrite(fo,out_buf,out,out_len);
                /* update checksum */
                if (checksum != null)
                    checksum.update(out_buf,out,out_len);
                total_d += out_len;
            }
            else
            {
                /* write original (incompressible) block */
                xwrite(fo,in_buf,in,in_len);
                /* update checksum */
                if (checksum != null)
                    checksum.update(in_buf,in,in_len);
                total_d += in_len;
            }

            total_c += in_len;
            total_blocks++;
        }

        /* read and verify checksum */
        if ((flags & 1) != 0) {
            long c = xread32(fi) & 0xffffffffL;
            if (checksum != null && c != checksum.getValue()) {
                throw new DataFormatException("checksum error - data corrupted");
            }
        }

        total_time += System.currentTimeMillis() - total_start_time;
    }


    //
    // stats
    //

    private void reset_stats() {
        total_in = 0;
        total_out = 0;
        total_blocks = 0;
        total_c = 0;
        total_d = 0;
        compress_time = 0;
        decompress_time = 0;
        total_time = 0;
    }

    void print_stats(String n) {
        PrintStream f = System.out;
        f.println("processing time: " + ((total_time) / 1000.0) + " seconds");
        if (compress_loops <= 1 && decompress_loops <= 1)
            return;

        float ratio = (total_d > 0) ? (100.0f * total_c) / total_d : 0;
        float c_kbs = 0;
        float d_kbs = 0;
        if (compress_time > 0.001)
            c_kbs = (1.0f * total_d * compress_loops) / compress_time;
        if (decompress_time > 0.001)
            d_kbs = (1.0f * total_d * decompress_loops) / decompress_time;

        if (decompress_time > 0.001)
            f.println("decompression speed: about " + d_kbs + " kB per second");

        // print a benchmark line for util/table.pl
        f.print("LZO1X [java] | " + n + " " + total_d);
        f.print(" " + total_blocks + " " + total_c);
        f.println(" " + ratio + " " + c_kbs + " " + d_kbs + " |");
    }


    //
    // main
    //

    static void usage(PrintStream f) {
        f.println("Usage: java lunpack -d [options] input-file output-file (decompress)");
        f.println("Usage: java lunpack -t [options] input-file...          (test)");
        f.println();
        f.println("Options:");
        f.println("  -F      do not compute a checksum (faster)");
        f.println("  -D#     run decompressor # times (benchmark)");
    }


    public static void main(String args[]) {
        int r = 0;
        int opt_block_size = 256 * 1024;
        boolean opt_decompress = false;
        boolean opt_test = false;
        String iname = null;
        String oname = null;
        FileInputStream fi = null;
        FileOutputStream fo = null;
        lunpack lp = new lunpack();
        lp.bc = new Lzo1xDecompressor();

        try {
            int i = 0;

            while (i < args.length && args[i].startsWith("-")) {
                if (args[i].equals("-d"))
                    opt_decompress = true;
                else if (args[i].equals("-t"))
                    opt_test = true;
                else if (args[i].equals("-F"))
                    lp.opt_checksum = false;
                else if (args[i].equals("--debug"))
                    lp.opt_debug = true;
                else if (args[i].startsWith("-C"))
                {
                    lp.compress_loops = 1;
                    String opt = args[i].substring(2);
                    int o = opt.length() > 0 ? Integer.parseInt(opt) : 0;
                    if (o > 0)
                        lp.compress_loops = o;
                }
                else if (args[i].startsWith("-D"))
                {
                    lp.decompress_loops = 1;
                    String opt = args[i].substring(2);
                    int o = opt.length() > 0 ? Integer.parseInt(opt) : 0;
                    if (o > 0)
                        lp.decompress_loops = o;
                }
                else if (args[i].startsWith("-b")) {
                    String opt = args[i].substring(2);
                    int o = opt.length() > 0 ? Integer.parseInt(opt) : 0;
                    if (o < 1024 || o > 1024*1024)
                        throw new IllegalArgumentException("invalid block size " + o);
                    opt_block_size = o;
                }
                else
                    throw new IllegalArgumentException("invalid option " + args[i]);
                i++;
            }

            if (opt_test) {
                if (i >= args.length)
                    throw new IllegalArgumentException();
                while (i < args.length) {
                    iname = args[i++];
                    fi = new FileInputStream(iname);
                    lp.do_decompress(fi,fo);
                    fi.close(); fi = null;
                    System.out.println(iname + ": tested ok (" + lp.total_in +
                                       " -> " + lp.total_out + " bytes)");
                    lp.print_stats(iname);
                }
            } else if (opt_decompress) {
                if (i + 2 != args.length)
                    throw new IllegalArgumentException();
                iname = args[i++];
                oname = args[i++];
                fi = new FileInputStream(iname);
                fo = new FileOutputStream(oname);
                lp.do_decompress(fi,fo);
                System.out.println("decompressed " + lp.total_in +
                                   " into " + lp.total_out + " bytes");
                lp.print_stats(iname);
            }
            else
                usage(System.err);
        } catch (IllegalArgumentException e) {
            r = 1;
            System.err.println(e);
            System.err.println();
            usage(System.err);
        } catch (Exception e) {
            r = 1;
            if (iname != null)
                System.err.print(iname + ": ");
            System.err.println(e);
        } finally {
            try {
                if (fi != null)
                    fi.close();
            } catch (Exception e) { }
            try {
                if (fo != null)
                    fo.close();
            } catch (Exception e) { }
        }

        System.exit(r);
    }
}


// vi:ts=4:et



java-lzo/org/lzo/Int.java
/* Int.java -- wraps the primitive type int

   This file is part of the LZO real-time data compression library.

   Copyright (C) 1999 Markus Franz Xaver Johannes Oberhumer
   Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer

   The LZO library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   The LZO library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the LZO library; see the file COPYING.
   If not, write to the Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Markus F.X.J. Oberhumer
   <markus.oberhumer@jk.uni-linz.ac.at>
   http://wildsau.idv.uni-linz.ac.at/mfx/lzo.html
 */


package org.lzo;


/***********************************************************************
 * The Int class wraps a value of the primititve type <code>int</code>
 * in an object. As opposed to <code>java.lang.Integer</code>
 * the value can be modified after the object has been created.
 *
 * @author  Markus F.X.J. Oberhumer <markus.oberhumer@jk.uni-linz.ac.at>
 * @see     java.lang.Integer
 ***********************************************************************/

public class Int
    extends java.lang.Number
    // implements java.io.Serializable      // @JDK@ 1.1
{
    private int value = 0;

    public Int() {
        this(0);
    }

    public Int(int v) {
        value = v;
    }

    public void setValue(int v) {
        value = v;
    }

    public void add(int v) {
        value += v;
    }

    public void sub(int v) {
        value -= v;
    }

    public byte byteValue() {
        return (byte) value;
    }

    public short shortValue() {
        return (short) value;
    }

    public int intValue() {
        return value;
    }

    public long longValue() {
        return value;
    }

    public float floatValue() {
        return value;
    }

    public double doubleValue() {
        return value;
    }
}


// vi:ts=4:et



java-lzo/org/lzo/DataFormatException.java
/* DataFormatException.java -- signals a data format error

   This file is part of the LZO real-time data compression library.

   Copyright (C) 1999 Markus Franz Xaver Johannes Oberhumer
   Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer

   The LZO library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   The LZO library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the LZO library; see the file COPYING.
   If not, write to the Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Markus F.X.J. Oberhumer
   <markus.oberhumer@jk.uni-linz.ac.at>
   http://wildsau.idv.uni-linz.ac.at/mfx/lzo.html
 */


package org.lzo;


/***********************************************************************
 * Signals that a data format error has occured.
 *
 * @author  Markus F.X.J. Oberhumer <markus.oberhumer@jk.uni-linz.ac.at>
 * @see     java.util.zip.DataFormatException
 ***********************************************************************/

public class DataFormatException
    extends Exception
{
    public DataFormatException() {
        super();
    }

    public DataFormatException(String s) {
        super(s);
    }

    public DataFormatException(int code) {
        this("error code " + code);
    }
}


// vi:ts=4:et



python-lzo/lzomodule.c
/* lzomodule.c -- LZO bindings for Python 1.5.1

   This file is part of the LZO real-time data compression library.

   Copyright (C) 1998 Markus Franz Xaver Johannes Oberhumer

   The LZO library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of
   the License, or (at your option) any later version.

   The LZO library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with the LZO library; see the file COPYING.
   If not, write to the Free Software Foundation, Inc.,
   59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   Markus F.X.J. Oberhumer
   <markus.oberhumer@jk.uni-linz.ac.at>
   http://wildsau.idv.uni-linz.ac.at/mfx/lzo.html
 */


#include "Python.h"

#if 0
#include <lzo/lzo1x.h>
#else
#include <lzo1x.h>
#endif

#if !defined(LZO_VERSION) || (LZO_VERSION < 0x1030)
#  error "Need LZO v1.03 or newer"
#endif

#define UNUSED(x)       x = x

static PyObject *LzoError;


/***********************************************************************
// compress
************************************************************************/

static /* const */ char compress__doc__[] =
"compress(string) -- Compress string using the default compression level, "
"returning a string containing compressed data.\n"
"compress(string, level) -- Compress string, using the chosen compression "
"level (from 1 to 9).  Return a string containing the compressed data.\n"
;

static PyObject *
compress(PyObject *self, PyObject *args)
{
	PyObject *result_str;
	const lzo_bytep in;
	lzo_bytep out;
	lzo_voidp wrkmem = NULL;
	lzo_uint in_len;
	lzo_uint out_len;
	lzo_uint new_len;
	int len;
	int err;
	int level = 1;

	/* init */
	UNUSED(self);
	if (!PyArg_ParseTuple(args, "s#|i", &in, &len, &level))
		return NULL;
	if (len < 0)
		return NULL;
	in_len = len;
	out_len = in_len + in_len / 64 + 16 + 3;

	/* alloc buffers */
	result_str = PyString_FromStringAndSize(NULL, 5 + out_len);
	if (level == 1)
		wrkmem = PyMem_Malloc(LZO1X_1_MEM_COMPRESS);
	else
		wrkmem = PyMem_Malloc(LZO1X_999_MEM_COMPRESS);
	if (result_str == NULL || wrkmem == NULL)
	{
		if (wrkmem) PyMem_Free(wrkmem);
		Py_XDECREF(result_str);
		PyErr_SetString(PyExc_MemoryError,
						"Can't allocate memory to compress data");
		return NULL;
	}

	/* compress */
	out = (lzo_bytep) PyString_AsString(result_str);
	new_len = out_len;
	if (level == 1)
	{
		out[0] = 0xf0;
		err = lzo1x_1_compress(in,in_len,out+5,&new_len,wrkmem);
	}
	else
	{
		out[0] = 0xf1;
		err = lzo1x_999_compress(in,in_len,out+5,&new_len,wrkmem);
	}
	PyMem_Free(wrkmem);
	if (err != LZO_E_OK || new_len > out_len)
	{
		Py_DECREF(result_str);
		PyErr_Format(LzoError, "Error %i while compressing data", err);
		return NULL;
	}

	/* return */
	out[1] = (in_len >> 24) & 0xff;
	out[2] = (in_len >> 16) & 0xff;
	out[3] = (in_len >>  8) & 0xff;
	out[4] = (in_len >>  0) & 0xff;
	if (_PyString_Resize(&result_str, 5 + new_len) == -1)
	{
        /* this should not happen as we are not enlarging the string */
		Py_XDECREF(result_str);
		PyErr_SetString(PyExc_MemoryError,
						"Out of memory while compressing data");
		return NULL;
	}
	return result_str;
}


/***********************************************************************
// decompress
************************************************************************/

static /* const */ char decompress__doc__[] =
"decompress(string) -- Decompress the data in string, returning a string containing the decompressed data.\n"
;

static PyObject *
decompress(PyObject *self, PyObject *args)
{
	PyObject *result_str;
	const lzo_bytep in;
	lzo_bytep out;
	lzo_uint in_len;
	lzo_uint out_len;
	lzo_uint new_len;
	int len;
	int err;

	/* init */
	UNUSED(self);
	if (!PyArg_ParseTuple(args, "s#", &in, &len))
		return NULL;
	if (len < 5 + 3 || in[0] < 0xf0 || in[0] > 0xf1)
	{
		PyErr_SetString(LzoError, "Invalid compressed data");
		return NULL;
	}
	in_len = len - 5;
	out_len = (in[1] << 24) | (in[2] << 16) | (in[3] << 8) | in[4];

	/* alloc buffers */
	result_str = PyString_FromStringAndSize(NULL,out_len);
	if (result_str == NULL)
	{
		PyErr_SetString(PyExc_MemoryError,
						"Can't allocate memory to decompress data");
		return NULL;
	}

	/* decompress */
	out = (lzo_bytep) PyString_AsString(result_str);
	new_len = out_len;
	err = lzo1x_decompress_safe(in+5,in_len,out,&new_len,NULL);
	if (err != LZO_E_OK || new_len != out_len)
	{
		Py_DECREF(result_str);
		PyErr_SetString(LzoError, "Compressed data violation");
		return NULL;
	}

	/* return */
	return result_str;
}


/***********************************************************************
// optimize
************************************************************************/

static /* const */ char optimize__doc__[] =
"optimize(string) -- Optimize the representation of the compressed data, returning a string containing the compressed data.\n"
;

static PyObject *
optimize(PyObject *self, PyObject *args)
{
	PyObject *result_str;
	lzo_bytep in;
	lzo_bytep out;
	lzo_uint in_len;
	lzo_uint out_len;
	lzo_uint new_len;
	int len;
	int err;

	/* init */
	UNUSED(self);
	if (!PyArg_ParseTuple(args, "s#", &in, &len))
		return NULL;
	if (len < 5 + 3 || in[0] < 0xf0 || in[0] > 0xf1)
	{
		PyErr_SetString(LzoError, "Invalid compressed data");
		return NULL;
	}
	in_len = len - 5;
	out_len = (in[1] << 24) | (in[2] << 16) | (in[3] << 8) | in[4];

	/* alloc buffers */
	result_str = PyString_FromStringAndSize(in,len);
	out = (lzo_bytep) PyMem_Malloc(out_len > 0 ? out_len : 1);
	if (result_str == NULL || out == NULL)
	{
		if (out) PyMem_Free(out);
		Py_XDECREF(result_str);
		PyErr_SetString(PyExc_MemoryError,
						"Can't allocate memory to optimize data");
		return NULL;
	}

	/* optimize */
	in = (lzo_bytep) PyString_AsString(result_str);
	new_len = out_len;
	err = lzo1x_optimize(in+5,in_len,out,&new_len,NULL);
	PyMem_Free(out);
	if (err != LZO_E_OK || new_len != out_len)
	{
		Py_DECREF(result_str);
		PyErr_SetString(LzoError, "Compressed data violation");
		return NULL;
	}

	/* return */
	return result_str;
}


/***********************************************************************
// adler32
************************************************************************/

static /* const */ char adler32__doc__[] =
"adler32(string) -- Compute an Adler-32 checksum of string, using "
"a default starting value, and returning an integer value.\n"
"adler32(string, value) -- Compute an Adler-32 checksum of string, using "
"the starting value provided, and returning an integer value\n"
;

static PyObject *
adler32(PyObject *self, PyObject *args)
{
	unsigned long val = lzo_adler32(0, NULL, 0);
	const lzo_bytep buf;
	int len;

	UNUSED(self);
	if (!PyArg_ParseTuple(args, "s#|l", &buf, &len, &val))
		return NULL;
	if (len < 0)
		return NULL;
	val = lzo_adler32((lzo_uint32)val, buf, len);
	return PyInt_FromLong(val);
}


/***********************************************************************
// crc32
************************************************************************/

static /* const */ char crc32__doc__[] =
"crc32(string) -- Compute a CRC-32 checksum of string, using "
"a default starting value, and returning an integer value.\n"
"crc32(string, value) -- Compute a CRC-32 checksum of string, using "
"the starting value provided, and returning an integer value.\n"
;

static PyObject *
crc32(PyObject *self, PyObject *args)
{
	unsigned long val = lzo_crc32(0, NULL, 0);
	const lzo_bytep buf;
	int len;

	UNUSED(self);
	if (!PyArg_ParseTuple(args, "s#|l", &buf, &len, &val))
		return NULL;
	if (len < 0)
		return NULL;
	val = lzo_crc32((lzo_uint32)val, buf, len);
	return PyInt_FromLong(val);
}


/***********************************************************************
// main
************************************************************************/

static /* const */ PyMethodDef methods[] =
{
	{"compress",   (PyCFunction)compress,   1, compress__doc__},
	{"decompress", (PyCFunction)decompress, 1, decompress__doc__},
	{"optimize",   (PyCFunction)optimize,   1, optimize__doc__},
	{"adler32",    (PyCFunction)adler32,    1, adler32__doc__},
	{"crc32",      (PyCFunction)crc32,      1, crc32__doc__},
	{NULL, NULL}
};


static /* const */ char module_documentation[]=
"The functions in this module allow compression and decompression "
"using the LZO library.\n\n"
"adler32(string) -- Compute an Adler-32 checksum.\n"
"adler32(string, start) -- Compute an Adler-32 checksum using a given starting value.\n"
"compress(string) -- Compress a string.\n"
"compress(string, level) -- Compress a string with the given level of compression (1-9).\n"
"crc32(string) -- Compute a CRC-32 checksum.\n"
"crc32(string, start) -- Compute a CRC-32 checksum using a given starting value.\n"
"decompress(string) -- Decompresses a compressed string.\n"
"optimize(string) -- Optimize a compressed string.\n"
;

void initlzo(void)
{
	PyObject *m, *d, *v;

	if (lzo_init() != LZO_E_OK)
		return;

	m = Py_InitModule4("lzo", methods, module_documentation,
					   NULL, PYTHON_API_VERSION);
	d = PyModule_GetDict(m);

	LzoError = PyErr_NewException("lzo.error", NULL, NULL);
	PyDict_SetItemString(d, "error", LzoError);

	v = PyInt_FromLong((long)lzo_version());
	PyDict_SetItemString(d, "LZO_VERSION", v);
	Py_DECREF(v);

	v = PyString_FromString(lzo_version_string());
	PyDict_SetItemString(d, "LZO_VERSION_STRING", v);
	Py_DECREF(v);

	v = PyString_FromString(lzo_version_date());
	PyDict_SetItemString(d, "LZO_VERSION_DATE", v);
	Py_DECREF(v);
}


/*
vi:ts=4
*/
